middle_end/flambda2/algorithms/patricia_tree.ml

```diff
@@ -848,212 +679,353 @@ struct
-        (Format.pp_print_list ~pp_sep:Format.pp_print_space
-          (fun ppf (key, datum) ->
-            Format.fprintf ppf "@[<hov 1>(%a@ %a)@]"
-              Key.print key print_datum datum))
-        (bindings t)
+  and 'a t0 = 'a Set0.t =
+    | Empty : unit t0
+    | Leaf : elt -> unit t0
+    | Branch : elt * elt * unit t0 * unit t0 -> unit t0
+
+  module Ops = Tree_operations (Set0)
+  include Ops
+
+  let empty = Empty
+
+  let singleton i = Ops.singleton Unit i ()
+
+  let add i t = Ops.add i () t
+
+  (* CR lmaurer: This is slow, but [Ops.union] is hard to specialize *)
+  let union t0 t1 = Ops.union (fun _ () () -> Some ()) t0 t1
+
+  let disjoint t0 t1 = not (Ops.inter_domain_is_non_empty t0 t1)
+
+  (* CR lmaurer: Also should probably be specialized *)
+  let inter t0 t1 = Ops.inter Unit (fun _ () () -> ()) t0 t1
+
+  let rec union_list ts =
+    match ts with [] -> empty | t :: ts -> union t (union_list ts)
+
+  let filter_map f t =
+    let rec loop f acc = function
+      | Empty -> acc
+      | Leaf i -> begin match f i with None -> acc | Some j -> add j acc end
+      | Branch (_, _, t0, t1) -> loop f (loop f acc t0) t1
+    in
+    loop f Empty t
+
+  let elements = Ops.to_list_unordered
+
+  let min_elt = Ops.min_binding
+
+  let min_elt_opt = Ops.min_binding_opt
+
+  let max_elt = Ops.max_binding
+
+  let max_elt_opt = Ops.max_binding_opt
+
+  let equal t0 t1 = Ops.equal (fun () () -> true) t0 t1
+
+  let compare t0 t1 = Ops.compare (fun () () -> 0) t0 t1
+
+  let split i t = Ops.split ~found:(fun () -> true) ~not_found:false i t
 
+  let find elt t = if mem elt t then elt else raise Not_found
+
+  let of_list l = Ops.of_list Unit l
+
+  let map f t = Ops.map_keys f t
+end
+
+module Map = struct
+  include Map0
+  module Ops = Tree_operations (Map0)
+  include Ops
+
+  let empty = Empty
+
+  let singleton i d = Ops.singleton Any i d
+
+  let inter f t0 t1 = Ops.inter Any f t0 t1
+
+  let split i t = Ops.split ~found:(fun a -> Some a) ~not_found:None i t
+
+  let bindings s =
+    List.sort
+      (fun (id1, _) (id2, _) -> Int.compare id1 id2)
+      (Ops.to_list_unordered s)
+
+  let map f t = Ops.map Any f t
+
+  let mapi f t = Ops.mapi Any f t
+
+  let filter_map f t =
+    fold
+      (fun id v map -> match f id v with None -> map | Some r -> add id r map)
+      t empty
+
+  let of_list l = Ops.of_list Any l
+
+  let merge f t0 t1 = Ops.merge Any f t0 t1
+
+  (* CR lmaurer: This should be doable as a map operation if we generalize
+     [Ops.map] by letting the returned tree be built by a different [Tree]
+     module *)
   let keys map = fold (fun k _ set -> Set.add k set) map Set.empty
 
   let data t = List.map snd (bindings t)
 
+  (* CR lmaurer: See comment on [keys] *)
   let of_set f set = Set.fold (fun e map -> add e (f e) map) set empty
 
   let diff_domains = diff
 
-  let () =
-    ignore subset;
-    ignore inter_domains;
-    ignore diff;
-    ignore elements
+  let () = ignore Ops.inter_domains
```

Comment:
Looks like `inter_domains` can be removed?

Reply:
(Maybe don't squash this branch and keep it around, so we can recover it later if we want it.)